greenpithumb/dht11.py

```diff
@@ -0,0 +1,50 @@
+import datetime
+import threading
+
+from Adafruit_DHT import common
+
+_FRESHNESS_THRESHOLD = 2
+
+
+class CachingDHT11(object):
+    """Caching DHT11 sensor wrapper.
+
+    Reads and returns temperature and humidity levels while also caching these
+    values to ensure that the sensor is not polled at too high of a frequency.
+    """
+
+    def __init__(self, dht11_pin, clock):
+        """Creates a new CachingDHT11 object.
+
+        Args:
+            dht11_pin: An int representing the pin that the DHT11 sensor is
+                connected to.
+            clock: A clock interface
+        """
+        self._dht11_pin = dht11_pin
+        self._clock = clock
+        self._last_reading_time = datetime.min
+        self._last_reading = None
+        self._lock = threading.Lock()
+
+    def humidity(self):
+        """Returns a recent relative humidity reading."""
+        self._lock.acquire()
```

Comment:
>Also, I think the lock is slightly too broad. It doesn't hurt anything, but just so we're thinking deliberately about it, I'm trying to work through what the lock needs to protect. The now = self._clock.now() can go outside the lock...

It seems to me like ```now``` should stay inside the lock because ```_last_reading_time``` should be coupled to ```_last_reading```. Eg. if ```now``` is outside the lock and it takes 1 second to acquire the lock, ```_last_reading``` will appear to have been taken 1 second before it actually was.